src/mono/wasm/runtime/invoke-js.ts

```diff
@@ -0,0 +1,316 @@
+    const covertor = cs_to_js_map.get(mono_type);
+    //console.log("covert_arg_to_js C " + typeof covertor);
+    assert(covertor, `Unknow covertor for type ${mono_type} at ${position}`);
+    //console.log("covert_arg_to_js D");
+    const r = covertor(<any>data + (position << 2));
+    //console.log("covert_arg_to_js E");
+    return r;
+}
+
+function covert_result_to_cs(signature: VoidPtr, data: VoidPtr, value: any): void {
+    //console.log("covert_result_to_cs A");
+    const mono_type = <MonoType><any>getU32(<any>signature);
+    //console.log("covert_result_to_cs B " + mono_type);
+    if (mono_type === MonoTypeNull) {
+        return;// void
+    }
+    console.log("covert_result_to_cs C");
+    const covertor = js_to_cs_map.get(mono_type);
+    console.log("covert_result_to_cs D");
+    assert(covertor, `Unknow covertor for type ${mono_type} at ret`);
+    console.log("covert_result_to_cs E");
+    covertor(<any>data, value);
+    console.log("covert_result_to_cs F");
+}
+
+function _marshal_int32_to_cs(target: VoidPtr, value: any) {
+    return setI32(target, value);
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function _marshal_void_to_cs(_target: VoidPtr, _value: any) {
+    return 0;
+}
+
+function _marshal_int32_to_js(source: VoidPtr) {
+    //console.log("_marshal_int32_to_js S" + source);
+    const pRef = getU32(source);
+    //console.log("_marshal_int32_to_js P" + pRef);
+    const value = getI32(pRef);
+    //console.log("_marshal_int32_to_js v" + value);
+    return value;
+}
+
+function _marshal_string_to_cs(target: VoidPtr, value: any) {
+    const pError = js_string_to_mono_string(value)!;
+    <MonoString><any>setU32(target, pError);
+}
+
+function _marshal_string_to_js(source: VoidPtr) {
+    //console.log("_marshal_string_to_js S" + source);
+    const pRef = getU32(source);
+    //console.log("_marshal_string_to_js P" + pRef);
+    const pValue = <MonoString><any>getU32(pRef);
+    //console.log("_marshal_string_to_js P2" + pValue);
+    const value = conv_string(pValue);
+    //console.log("_marshal_string_to_js v" + value);
+    return value;
+}
+
+function _marshal_exception_to_cs(target: VoidPtr, value: any) {
+    let res = "unknown exception";
+    if (value) {
+        res = value.toString();
+        const stack = value.stack;
+        if (stack) {
+            // Some JS runtimes insert the error message at the top of the stack, some don't,
+            //  so normalize it by using the stack as the result if it already contains the error
+            if (stack.startsWith(res))
+                res = stack;
+            else
+                res += "\n" + stack;
+        }
+    }
+
+    const pError = js_string_to_mono_string(res)!;
+    setU32(target, pError);
+}
+
+function _marshal_exception_to_js(source: VoidPtr) {
+    const pValue = <MonoString><any>getU32(source);
+    const jsValue = conv_string(pValue);
+    return new Error(jsValue!);
```

Comment:
Later on we should consider doing something here to transport the error message and stack separately in both directions so that you don't get a massive message with multiple stacks mixed in when you transition between languages. In JS we should be able to override the ```stack``` property with a custom one to do a mixed-language stack, and in C# we can store the JS stack in its own property next to the message.

Reply:
Added TODO comment for now.